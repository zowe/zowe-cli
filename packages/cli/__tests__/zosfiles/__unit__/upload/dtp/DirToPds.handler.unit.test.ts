/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Upload } from "@zowe/zos-files-for-zowe-sdk";
import { UNIT_TEST_ZOSMF_PROF_OPTS } from "../../../../../../../__tests__/__src__/mocks/ZosmfProfileMock";
import DirToPdsHandler from "../../../../../src/zosfiles/upload/dtp/DirToPds.handler";

describe("Upload dir-to-pds handler", () => {
    describe("process method", () => {
        let handler: DirToPdsHandler;
        beforeEach(() => {
            jest.resetAllMocks();
            handler = new DirToPdsHandler();
            jest.mock('../../../../../src/zosfiles/upload/dtp/DirToPds.handler', () => ({
                checkDirectoryExistence: jest.fn().mockResolvedValue(undefined)
            }));
        });

        it("should upload a directory to a PDS if requested", async () => {
            const inputdir = "test-dir";
            const dataSetName = "testing";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession = null;

            // Mock the submit JCL function
            Upload.dirToPds = jest.fn(async (session) => {
                fakeSession = session;
                return {
                    success: false,
                    commandResponse: "uploaded",
                    apiResponse: [
                        {success: true, from: inputdir, to: dataSetName},
                        {success: false, from: "testfrom", to: "testto"},
                        {success: undefined, from: "dummy", to: "nowhere"}
                    ]
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.process({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        inputdir,
                        dataSetName,
                        ...UNIT_TEST_ZOSMF_PROF_OPTS
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        },
                        progress: {
                            startBar: jest.fn((parms) => {
                                // do nothing
                            }),
                            endBar: jest.fn(() => {
                                // do nothing
                            })
                        }
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(Upload.dirToPds).toHaveBeenCalledTimes(1);
            expect(Upload.dirToPds).toHaveBeenCalledWith(fakeSession, inputdir, dataSetName, {
                task: {
                    percentComplete: 0,
                    stageName: 0,
                    statusMessage: "Uploading directory to PDS"
                }
            });
            expect(jsonObj).toMatchSnapshot();
            expect(apiMessage).toMatchSnapshot();
            expect(logMessage).toMatchSnapshot();
        });

        // it("shouldn't attempt to upload a file that hasnt been found", async () => {

        // })

        it("should upload a directory to a PDS in binary format if requested", async () => {
            const inputdir = "test-dir";
            const dataSetName = "testing";
            const binary = true;

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession = null;

            // Mock the submit JCL function
            Upload.dirToPds = jest.fn(async (session) => {
                fakeSession = session;
                return {
                    success: false,
                    commandResponse: "uploaded",
                    apiResponse: [
                        {success: true, from: inputdir, to: dataSetName},
                        {success: false, from: "testfrom", to: "testto"},
                        {success: undefined, from: "dummy", to: "nowhere"}
                    ]
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.process({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        inputdir,
                        dataSetName,
                        binary,
                        ...UNIT_TEST_ZOSMF_PROF_OPTS
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        },
                        progress: {
                            startBar: jest.fn((parms) => {
                                // do nothing
                            }),
                            endBar: jest.fn(() => {
                                // do nothing
                            })
                        }
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(Upload.dirToPds).toHaveBeenCalledTimes(1);
            expect(Upload.dirToPds).toHaveBeenCalledWith(fakeSession, inputdir, dataSetName, {
                binary,
                task: {
                    percentComplete: 0,
                    stageName: 0,
                    statusMessage: "Uploading directory to PDS"
                }
            });
            expect(jsonObj).toMatchSnapshot();
            expect(apiMessage).toMatchSnapshot();
            expect(logMessage).toMatchSnapshot();
        });

        it("should upload a directory to a PDS in record format if requested", async () => {
            const inputdir = "test-dir";
            const dataSetName = "testing";
            const record = true;

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession = null;

            // Mock the submit JCL function
            Upload.dirToPds = jest.fn(async (session) => {
                fakeSession = session;
                return {
                    success: false,
                    commandResponse: "uploaded",
                    apiResponse: [
                        {success: true, from: inputdir, to: dataSetName},
                        {success: false, from: "testfrom", to: "testto"},
                        {success: undefined, from: "dummy", to: "nowhere"}
                    ]
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.process({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        inputdir,
                        dataSetName,
                        record,
                        ...UNIT_TEST_ZOSMF_PROF_OPTS
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        },
                        progress: {
                            startBar: jest.fn((parms) => {
                                // do nothing
                            }),
                            endBar: jest.fn(() => {
                                // do nothing
                            })
                        }
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(Upload.dirToPds).toHaveBeenCalledTimes(1);
            expect(Upload.dirToPds).toHaveBeenCalledWith(fakeSession, inputdir, dataSetName, {
                record,
                task: {
                    percentComplete: 0,
                    stageName: 0,
                    statusMessage: "Uploading directory to PDS"
                }
            });
            expect(jsonObj).toMatchSnapshot();
            expect(apiMessage).toMatchSnapshot();
            expect(logMessage).toMatchSnapshot();
        });
    });
});
