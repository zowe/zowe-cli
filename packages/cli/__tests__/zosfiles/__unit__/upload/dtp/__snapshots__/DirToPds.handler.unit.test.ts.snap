// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload dir-to-pds handler process method should not attempt to upload a dir that hasn't been found 1`] = `undefined`;

exports[`Upload dir-to-pds handler process method should not attempt to upload a dir that hasn't been found 2`] = `""`;

exports[`Upload dir-to-pds handler process method should not attempt to upload a dir that hasn't been found 3`] = `
"
Error: Failed to access directory"
`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS if requested 1`] = `
Object {
  "apiResponse": Array [
    Object {
      "from": "test-dir",
      "success": true,
      "to": "testing",
    },
    Object {
      "from": "testfrom",
      "success": false,
      "to": "testto",
    },
    Object {
      "from": "dummy",
      "success": undefined,
      "to": "nowhere",
    },
  ],
  "commandResponse": "uploaded",
  "success": false,
}
`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS if requested 2`] = `""`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS if requested 3`] = `
"
uploaded"
`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS in binary format if requested 1`] = `
Object {
  "apiResponse": Array [
    Object {
      "from": "test-dir",
      "success": true,
      "to": "testing",
    },
    Object {
      "from": "testfrom",
      "success": false,
      "to": "testto",
    },
    Object {
      "from": "dummy",
      "success": undefined,
      "to": "nowhere",
    },
  ],
  "commandResponse": "uploaded",
  "success": false,
}
`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS in binary format if requested 2`] = `""`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS in binary format if requested 3`] = `
"
uploaded"
`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS in record format if requested 1`] = `
Object {
  "apiResponse": Array [
    Object {
      "from": "test-dir",
      "success": true,
      "to": "testing",
    },
    Object {
      "from": "testfrom",
      "success": false,
      "to": "testto",
    },
    Object {
      "from": "dummy",
      "success": undefined,
      "to": "nowhere",
    },
  ],
  "commandResponse": "uploaded",
  "success": false,
}
`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS in record format if requested 2`] = `""`;

exports[`Upload dir-to-pds handler process method should upload a directory to a PDS in record format if requested 3`] = `
"
uploaded"
`;
